File: desktop/verium-app/src-tauri/crates/verium-pool-miner/src/protocol.rs
//! Share reconstruction and PoW target math (port of `@pool/protocol`).

use num_bigint::BigUint;
use sha2::{Digest, Sha256};

pub const SCRYPT2_STRATUM_FACTOR: f64 = 65536.0;

const DIFF1_HEX: &str = "00000000ffff0000000000000000000000000000000000000000000000000000";
const DIFF_PRECISION: u64 = 100_000_000;

/// Double SHA-256.
pub fn sha256d(buf: &[u8]) -> [u8; 32] {
    let a = Sha256::digest(buf);
    Sha256::digest(a).into()
}

pub fn reverse_buffer(buf: &[u8]) -> Vec<u8> {
    buf.iter().rev().copied().collect()
}

fn reverse_to_32(buf: &[u8]) -> [u8; 32] {
    let mut out = [0u8; 32];
    let len = buf.len().min(32);
    for i in 0..len {
        out[i] = buf[buf.len() - 1 - i];
    }
    out
}

fn decode_prev_internal(prev_hash_hex: &str) -> Result<[u8; 32], String> {
    let prev = hex::decode(prev_hash_hex).map_err(|e| format!("prev_hash hex: {e}"))?;
    Ok(reverse_to_32(&prev))
}

pub fn stratum_to_effective_difficulty(stratum_diff: f64) -> f64 {
    stratum_diff / SCRYPT2_STRATUM_FACTOR
}

fn diff1_target() -> BigUint {
    BigUint::parse_bytes(DIFF1_HEX.as_bytes(), 16).expect("DIFF1")
}

fn max_u256() -> BigUint {
    (BigUint::from(1u8) << 256) - 1u8
}

fn biguint_to_be32(v: BigUint) -> [u8; 32] {
    let bytes = v.to_bytes_be();
    let mut out = [0u8; 32];
    let start = 32usize.saturating_sub(bytes.len());
    out[start..].copy_from_slice(&bytes);
    out
}

/// Convert fractional difficulty to a 256-bit target.
pub fn difficulty_to_target(difficulty: f64) -> [u8; 32] {
    if difficulty <= 0.0 {
        return biguint_to_be32(max_u256());
    }
    let scaled = (difficulty * DIFF_PRECISION as f64).round() as u64;
    if scaled == 0 {
        return biguint_to_be32(max_u256());
    }
    let target = diff1_target() * BigUint::from(DIFF_PRECISION) / BigUint::from(scaled);
    let capped = if target > max_u256() {
        max_u256()
    } else {
        target
    };
    biguint_to_be32(capped)
}

/// Interpret a 32-byte scrypt² hash as a 256-bit integer (reverse bytes, big-endian read).
pub fn hash_to_biguint(hash: &[u8; 32]) -> BigUint {
    let mut rev = *hash;
    rev.reverse();
    BigUint::from_bytes_be(&rev)
}

pub fn hash_meets_target(hash: &[u8; 32], target: &[u8; 32]) -> bool {
    hash_meets_target_fast(hash, target)
}

/// Allocation-free target check (consensus hash byte order vs big-endian target).
pub fn hash_meets_target_fast(hash: &[u8; 32], target: &[u8; 32]) -> bool {
    let mut rev = *hash;
    rev.reverse();
    rev <= *target
}

pub fn merkle_join(a: &[u8; 32], b: &[u8; 32]) -> [u8; 32] {
    let mut concat = [0u8; 64];
    concat[..32].copy_from_slice(a);
    concat[32..].copy_from_slice(b);
    sha256d(&concat)
}

pub fn merkle_root_with_coinbase(coinbase_hash: [u8; 32], steps: &[[u8; 32]]) -> [u8; 32] {
    let mut root = coinbase_hash;
    for step in steps {
        root = merkle_join(&root, step);
    }
    root
}

#[derive(Clone, Debug)]
pub struct StratumJob {
    pub job_id: String,
    /// GBT display-order previous block hash (hex).
    pub prev_hash_hex: String,
    pub coinb1_hex: String,
    pub coinb2_hex: String,
    pub merkle_steps_hex: Vec<String>,
    pub version: u32,
    pub bits_hex: String,
    pub ntime: u32,
    pub ntime_hex: String,
}

#[derive(Clone, Debug)]
pub struct ShareSubmission {
    pub extranonce1: Vec<u8>,
    pub extranonce2_hex: String,
    pub ntime: u32,
    pub nonce: u32,
}

#[derive(Clone, Debug)]
pub struct ReconstructedShare {
    pub consensus_header: [u8; 80],
    pub merkle_root: [u8; 32],
}

/// Per-worker header template: merkle and fixed fields precomputed; only nonce changes per hash.
#[derive(Clone, Debug)]
pub struct WorkerHeaderWork {
    pub job_id: String,
    pub ntime_hex: String,
    pub extranonce2_hex: String,
    pub header_base: [u8; 80],
}

impl WorkerHeaderWork {
    pub fn header_with_nonce(&self, nonce: u32) -> [u8; 80] {
        let mut header = self.header_base;
        header[76..80].copy_from_slice(&nonce.to_le_bytes());
        header
    }
}

pub fn assemble_coinbase(
    coinb1: &[u8],
    extranonce1: &[u8],
    extranonce2: &[u8],
    coinb2: &[u8],
) -> Vec<u8> {
    let mut out =
        Vec::with_capacity(coinb1.len() + extranonce1.len() + extranonce2.len() + coinb2.len());
    out.extend_from_slice(coinb1);
    out.extend_from_slice(extranonce1);
    out.extend_from_slice(extranonce2);
    out.extend_from_slice(coinb2);
    out
}

/// Consensus 80-byte header (veriumd / pool share validation).
pub fn serialize_header(
    version: u32,
    prev_hash_hex: &str,
    merkle_root: &[u8; 32],
    ntime: u32,
    bits_hex: &str,
    nonce: u32,
) -> [u8; 80] {
    let prev_internal = decode_prev_internal(prev_hash_hex).expect("prev_hash hex");
    serialize_header_parts(version, &prev_internal, merkle_root, ntime, bits_hex, nonce)
}

fn serialize_header_parts(
    version: u32,
    prev_internal: &[u8; 32],
    merkle_root: &[u8; 32],
    ntime: u32,
    bits_hex: &str,
    nonce: u32,
) -> [u8; 80] {
    let bits = u32::from_str_radix(bits_hex, 16).unwrap_or(0);
    let mut header = [0u8; 80];
    header[0..4].copy_from_slice(&version.to_le_bytes());
    header[4..36].copy_from_slice(prev_internal);
    header[36..68].copy_from_slice(merkle_root);
    header[68..72].copy_from_slice(&ntime.to_le_bytes());
    header[72..76].copy_from_slice(&bits.to_le_bytes());
    header[76..80].copy_from_slice(&nonce.to_le_bytes());
    header
}

/// Build a worker-local header template (nonce zero) for the mining hot loop.
pub fn prepare_worker_header_work(
    job: &StratumJob,
    extranonce1: &[u8],
    extranonce2: &[u8],
) -> Result<WorkerHeaderWork, String> {
    let coinb1 = hex::decode(&job.coinb1_hex).map_err(|e| format!("coinb1: {e}"))?;
    let coinb2 = hex::decode(&job.coinb2_hex).map_err(|e| format!("coinb2: {e}"))?;
    let mut coinbase = assemble_coinbase(&coinb1, extranonce1, extranonce2, &coinb2);
    if coinbase.len() >= 8 {
        coinbase[4..8].copy_from_slice(&job.ntime.to_le_bytes());
    }
    let coinbase_hash = sha256d(&coinbase);
    let mut steps = Vec::with_capacity(job.merkle_steps_hex.len());
    for h in &job.merkle_steps_hex {
        let decoded = hex::decode(h).map_err(|e| format!("merkle step: {e}"))?;
        let mut b = [0u8; 32];
        let len = decoded.len().min(32);
        b[..len].copy_from_slice(&decoded[..len]);
        steps.push(b);
    }
    let merkle_root = merkle_root_with_coinbase(coinbase_hash, &steps);
    let prev_internal = decode_prev_internal(&job.prev_hash_hex)?;
    let header_base = serialize_header_parts(
        job.version,
        &prev_internal,
        &merkle_root,
        job.ntime,
        &job.bits_hex,
        0,
    );
    Ok(WorkerHeaderWork {
        job_id: job.job_id.clone(),
        ntime_hex: job.ntime_hex.clone(),
        extranonce2_hex: hex::encode(extranonce2),
        header_base,
    })
}

pub fn reconstruct_share(job: &StratumJob, sub: &ShareSubmission) -> ReconstructedShare {
    let en2 = hex::decode(&sub.extranonce2_hex).expect("extranonce2");
    let work = prepare_worker_header_work(job, &sub.extranonce1, &en2).expect("prepare work");
    let consensus_header = work.header_with_nonce(sub.nonce);
    let merkle_root = {
        let mut root = [0u8; 32];
        root.copy_from_slice(&consensus_header[36..68]);
        root
    };
    ReconstructedShare {
        consensus_header,
        merkle_root,
    }
}

pub fn uint32_le_hex(n: u32) -> String {
    hex::encode(n.to_le_bytes())
}

/// Parse `stratum+tcp://host:port` or `host:port`.
pub fn parse_stratum_url(url: &str) -> Result<(String, u16), String> {
    let trimmed = url.trim();
    let without_scheme = trimmed
        .strip_prefix("stratum+tcp://")
        .or_else(|| trimmed.strip_prefix("stratum://"))
        .unwrap_or(trimmed);
    let (host, port_str) = without_scheme
        .rsplit_once(':')
        .ok_or_else(|| format!("invalid stratum url: {url}"))?;
    let port: u16 = port_str
        .parse()
        .map_err(|_| format!("invalid stratum port: {port_str}"))?;
    Ok((host.to_string(), port))
}

File: src/stratum/protocol.cpp
// Copyright (c) 2026 The Vericonomy developers
// Distributed under the MIT software license, see the accompanying
// file COPYING or http://www.opensource.org/licenses/mit-license.php.

#include <stratum/protocol.h>

#include <arith_uint256.h>
#include <hash.h>
#include <util/strencodings.h>

#include <algorithm>
#include <cstring>

namespace stratum {
namespace {

constexpr uint64_t DIFF_PRECISION = 100000000ULL;

static std::string TrimCopy(const std::string& s)
{
    const size_t start = s.find_first_not_of(" \t\r\n");
    if (start == std::string::npos) {
        return {};
    }
    const size_t end = s.find_last_not_of(" \t\r\n");
    return s.substr(start, end - start + 1);
}

void Sha256d(const uint8_t* data, size_t len, uint8_t out[32])
{
    CHash256 hasher;
    hasher.Write(data, len).Finalize(out);
}

void ReverseTo32(const std::vector<unsigned char>& buf, uint8_t out[32])
{
    memset(out, 0, 32);
    const size_t len = std::min<size_t>(buf.size(), 32);
    for (size_t i = 0; i < len; ++i) {
        out[i] = buf[buf.size() - 1 - i];
    }
}

bool DecodePrevInternal(const std::string& prev_hash_hex, uint8_t out[32])
{
    const std::vector<unsigned char> prev = ParseHex(prev_hash_hex);
    if (prev.empty()) {
        return false;
    }
    ReverseTo32(prev, out);
    return true;
}

void SerializeHeaderParts(
    uint32_t version,
    const uint8_t prev_internal[32],
    const uint8_t merkle_root[32],
    uint32_t ntime,
    const std::string& bits_hex,
    uint32_t nonce,
    uint8_t header[80])
{
    uint32_t bits = 0;
    if (!bits_hex.empty()) {
        bits = (uint32_t)strtoul(bits_hex.c_str(), nullptr, 16);
    }
    memcpy(header + 0, &version, 4);
    memcpy(header + 4, prev_internal, 32);
    memcpy(header + 36, merkle_root, 32);
    memcpy(header + 68, &ntime, 4);
    memcpy(header + 72, &bits, 4);
    memcpy(header + 76, &nonce, 4);
}

uint8_t MerkleJoin(const uint8_t a[32], const uint8_t b[32], uint8_t out[32])
{
    uint8_t concat[64];
    memcpy(concat, a, 32);
    memcpy(concat + 32, b, 32);
    Sha256d(concat, 64, out);
    return 0;
}

void MerkleRootWithCoinbase(const uint8_t coinbase_hash[32], const std::vector<std::vector<unsigned char>>& steps, uint8_t out[32])
{
    memcpy(out, coinbase_hash, 32);
    for (const auto& step : steps) {
        uint8_t step32[32]{};
        const size_t len = std::min<size_t>(step.size(), 32);
        memcpy(step32, step.data(), len);
        MerkleJoin(out, step32, out);
    }
}

std::vector<uint8_t> AssembleCoinbase(
    const std::vector<unsigned char>& coinb1,
    const std::vector<uint8_t>& extranonce1,
    const std::vector<uint8_t>& extranonce2,
    const std::vector<unsigned char>& coinb2)
{
    std::vector<uint8_t> out;
    out.reserve(coinb1.size() + extranonce1.size() + extranonce2.size() + coinb2.size());
    out.insert(out.end(), coinb1.begin(), coinb1.end());
    out.insert(out.end(), extranonce1.begin(), extranonce1.end());
    out.insert(out.end(), extranonce2.begin(), extranonce2.end());
    out.insert(out.end(), coinb2.begin(), coinb2.end());
    return out;
}

void ArithToTargetBE(const arith_uint256& bn, uint8_t target[32])
{
    std::string hex = bn.GetHex();
    while (hex.size() < 64) {
        hex = "0" + hex;
    }
    const std::vector<unsigned char> bytes = ParseHex(hex);
    memset(target, 0, 32);
    const size_t len = std::min<size_t>(bytes.size(), 32);
    memcpy(target + (32 - len), bytes.data() + (bytes.size() - len), len);
}

} // namespace

double StratumToEffectiveDifficulty(double wire_difficulty)
{
    return wire_difficulty / SCRYPT2_STRATUM_FACTOR;
}

void DifficultyToTargetWords(double difficulty, uint32_t target[8])
{
    if (difficulty <= 0.0) {
        memset(target, 0xff, 32);
        return;
    }
    int k = 6;
    double diff = difficulty;
    while (k > 0 && diff > 1.0) {
        diff /= 4294967296.0;
        --k;
    }
    const uint64_t m = (uint64_t)(4294901760.0 / diff);
    if (m == 0 && k == 6) {
        memset(target, 0xff, 32);
        return;
    }
    memset(target, 0, 32);
    target[k] = (uint32_t)m;
    target[k + 1] = (uint32_t)(m >> 32);
}

void DifficultyToTarget(double difficulty, uint8_t target[32])
{
    if (difficulty <= 0.0) {
        memset(target, 0xff, 32);
        return;
    }
    const uint64_t scaled = (uint64_t)(difficulty * (double)DIFF_PRECISION + 0.5);
    if (scaled == 0) {
        memset(target, 0xff, 32);
        return;
    }
    arith_uint256 diff1;
    diff1.SetHex("00000000ffff0000000000000000000000000000000000000000000000000000");
    arith_uint256 bnTarget = diff1 * arith_uint256(DIFF_PRECISION) / arith_uint256(scaled);
    arith_uint256 maxUint;
    maxUint.SetHex("ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff");
    if (bnTarget > maxUint) {
        bnTarget = maxUint;
    }
    ArithToTargetBE(bnTarget, target);
}

bool HashMeetsTarget(const uint8_t hash[32], const uint8_t target[32])
{
    uint8_t rev[32];
    for (int i = 0; i < 32; ++i) {
        rev[i] = hash[31 - i];
    }
    return memcmp(rev, target, 32) <= 0;
}

bool ParseStratumUrl(const std::string& url, std::string& host_out, uint16_t& port_out)
{
    std::string trimmed = TrimCopy(url);
    const std::string prefixes[] = {"stratum+tcp://", "stratum://"};
    for (const auto& prefix : prefixes) {
        if (trimmed.rfind(prefix, 0) == 0) {
            trimmed = trimmed.substr(prefix.size());
            break;
        }
    }
    const size_t colon = trimmed.rfind(':');
    if (colon == std::string::npos || colon == 0 || colon + 1 >= trimmed.size()) {
        return false;
    }
    host_out = trimmed.substr(0, colon);
    try {
        const int p = std::stoi(trimmed.substr(colon + 1));
        if (p <= 0 || p > 65535) {
            return false;
        }
        port_out = (uint16_t)p;
    } catch (...) {
        return false;
    }
    return !host_out.empty();
}

bool PrepareWorkerHeaderWork(
    const StratumJob& job,
    const std::vector<uint8_t>& extranonce1,
    const std::vector<uint8_t>& extranonce2,
    WorkerHeaderWork& work_out,
    std::string& error_out)
{
    const std::vector<unsigned char> coinb1 = ParseHex(job.coinb1_hex);
    const std::vector<unsigned char> coinb2 = ParseHex(job.coinb2_hex);
    if (coinb1.empty() && !job.coinb1_hex.empty()) {
        error_out = "invalid coinb1";
        return false;
    }
    if (coinb2.empty() && !job.coinb2_hex.empty()) {
        error_out = "invalid coinb2";
        return false;
    }

    std::vector<uint8_t> coinbase = AssembleCoinbase(coinb1, extranonce1, extranonce2, coinb2);
    if (coinbase.size() >= 8) {
        memcpy(coinbase.data() + 4, &job.ntime, 4);
    }
    uint8_t coinbase_hash[32];
    Sha256d(coinbase.data(), coinbase.size(), coinbase_hash);

    std::vector<std::vector<unsigned char>> steps;
    steps.reserve(job.merkle_steps_hex.size());
    for (const auto& h : job.merkle_steps_hex) {
        steps.push_back(ParseHex(h));
    }
    uint8_t merkle_root[32];
    MerkleRootWithCoinbase(coinbase_hash, steps, merkle_root);

    uint8_t prev_internal[32];
    if (!DecodePrevInternal(job.prev_hash_hex, prev_internal)) {
        error_out = "invalid prev_hash";
        return false;
    }

    work_out.job_id = job.job_id;
    work_out.ntime_hex = job.ntime_hex;
    work_out.extranonce2_hex = HexStr(extranonce2);
    SerializeHeaderParts(job.version, prev_internal, merkle_root, job.ntime, job.bits_hex, 0, work_out.header_base);
    return true;
}

void HeaderWithNonce(const WorkerHeaderWork& work, uint32_t nonce, uint8_t header[80])
{
    memcpy(header, work.header_base, 80);
    memcpy(header + 76, &nonce, 4);
}

std::string Uint32LeHex(uint32_t n)
{
    uint8_t b[4];
    memcpy(b, &n, 4);
    return HexStr(b, b + 4);
}

} // namespace stratum
